File: leetcode-problems/538_convert_bst_to_greater_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def dfsBST(self, root, sum_so_far=0):
        if not root:
            return sum_so_far
        root.val += self.dfsBST(root.right, sum_so_far)
        return self.dfsBST(root.left, root.val)

    def convertBST(self, root):
        self.dfsBST(root)
        return root

File: leetcode/java/538ConvertBSTtoGreaterTree.java
package leetcode.java;

/**
 * T: O(N) S: O(N)
 *
 * Find the sum of all of the nodes in right subtree and add it to root. Now pass the value of the
 * root to the left subtree, as root and it's right subtree are the values that are guaranteed to be
 * greater. Add all of the values of the nodes in the left subtree to this sum, and return it to the
 * previous caller since it's the total sum of the subtree starting at root.
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {
    public TreeNode convertBST(TreeNode root) {
        calculateGreaterTree(root, 0);
        return root;
    }

    public int calculateGreaterTree(TreeNode root, int totalSoFar) {
        if (root == null) {
            return totalSoFar;
        }
        root.val += calculateGreaterTree(root.right, totalSoFar);
        return calculateGreaterTree(root.left, root.val);
    }
}
